### PR TITLE
Bug 1715635: docs/user/aws/install_upi: Document bring-your-own-DNS

### PR DESCRIPTION
**Pull request by wking**
Some users [want to provide their own `*.apps` DNS records][1] instead of delegating that to the ingress operator [1].  With this commit, we tell the ingress operator not to worry about managing any hosted zones, and walk users through how they can create [the expected records][2] themselves.

This is a WIP until I fill in a few outstanding FIXMEs.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1715635
[2]: https://github.com/openshift/cluster-ingress-operator/blob/9ce86811e6aeea58d3b2b8955591d50cd9311885/pkg/operator/controller/ingress/dns.go#L75-L115

### CODE REVIEW

**Comment by wking** [comment]
I've pushed a97308994 -> d70f94bef addressing the record-creation FIXME.  I've dropped the DNSRecord-monitoring FIXME for now, although we might want to add that later (example DNSRecord and note to this effect in the d70f94bef commit message).

**Comment by wking** [comment]
[e2e-aws-upgrade][1]:

```
2019/08/23 01:28:30 problem gathering successful build openstack-installer logs into artifacts: error: Unable to retrieve logs for build openstack-installer: pods "openstack-installer-build" not found
```

/retest

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/2221/pull-ci-openshift-installer-master-e2e-aws-upgrade/1946#0:build-log.txt%3A27

**Comment by wking** [comment]
Green enough.  CC @abhinavdahiya for review.

File: docs/user/aws/install_upi.md [hunk 1]
```diff
@@ -20,26 +18,59 @@ $ openshift-install create install-config
 ? Pull Secret [? for help]
 ```
 
-Edit the resulting `openshift-install.yaml` to set `replicas` to 0 for the `compute` pool:
+### Empty Compute Pools
 
-```console
-$ sed -i '1,/replicas: / s/replicas: .*/replicas: 0/' install-config.yaml
+We'll be providing the control-plane and compute machines ourselves, so edit the resulting `openshift-install.yaml` to set `replicas` to 0 for the `compute` pool:
```
Comment (abhinavdahiya) [inline docs/user/aws/install_upi.md]: install-config.yaml

File: docs/user/aws/install_upi.md [hunk 2]
```diff
@@ -20,26 +18,59 @@ $ openshift-install create install-config
...
 
 ```console
 $ rm -f openshift/99_openshift-cluster-api_master-machines-*.yaml openshift/99_openshift-cluster-api_worker-machinesets-*.yaml
 ```
 
 You are free to leave the compute MachineSets in if you want to create compute machines via the machine API, but if you do you may need to update the various references (`subnet`, etc.) to match your environment.
+
+### Remove DNS Zones
+
+If you don't want the ingress operator to create DNS records on your behalf, remove the `privateZone` and `publicZone` sections from the DNS configuration:
```
Comment (abhinavdahiya) [inline docs/user/aws/install_upi.md]: We linked machine-api operator in previous section

File: docs/user/aws/install_upi.md [hunk 3]
```diff
@@ -20,26 +18,59 @@ $ openshift-install create install-config
...
+path = "manifests/cluster-dns-02-config.yml";
+data = yaml.load(open(path));
+del data["spec"]["publicZone"];
+del data["spec"]["privateZone"];
+open(path, "w").write(yaml.dump(data, default_flow_style=False))'
+```
+
+If you do so, you'll need to [add](#add-the-ingress-dns-records) the following records to your private and public zones:
+
+* `*.apps.{baseDomain}.`, which should point at the ingress load balancer.
```
Comment (abhinavdahiya) [inline docs/user/aws/install_upi.md]: I think we were also going to list the required subdomains if the user didn't want to put wildcard and mention that each new route will have to manually added later on.

File: docs/user/aws/install_upi.md [hunk 4]
```diff
@@ -241,6 +272,61 @@ openshift-service-catalog-apiserver-operator            openshift-service-catalo
...
+```console
+$ oc -n openshift-ingress get service router-default
+NAME             TYPE           CLUSTER-IP      EXTERNAL-IP                                                              PORT(S)                      AGE
+router-default   LoadBalancer   172.30.62.215   ab37f072ec51d11e98a7a02ae97362dd-240922428.us-east-2.elb.amazonaws.com   80:31499/TCP,443:30693/TCP   5m
+```
+
+Find the hosted zone ID for the load balancer (or use [this table][route53-zones-for-load-balancers]):
+
+```console
+$ aws elb describe-load-balancers | jq -r '.LoadBalancerDescriptions[] | select(.DNSName == "ab37f072ec51d11e98a7a02ae97362dd-240922428.us-east-2.elb.amazonaws.com").CanonicalHostedZoneNameID'
```
Comment (abhinavdahiya) [inline docs/user/aws/install_upi.md]: Do the docs already use jq somewhere else?

Comment (wking) [reply docs/user/aws/install_upi.md]: > Do the docs already use jq somewhere else?

Yup:

```console
$ git grep jq origin/master -- docs/user/aws/install_upi.md
origin/master:docs/user/aws/install_upi.md:$ jq -r .infraID metadata.json
```

File: docs/user/aws/install_upi.md (hunk 1 above)
Comment (wking) [reply docs/user/aws/install_upi.md]: > install-config.yaml

Oops.  Fixed in d70f94bef -> 65cbc79e3.

File: docs/user/aws/install_upi.md (hunk 2 above)
Comment (wking) [reply docs/user/aws/install_upi.md]: > We linked machine-api operator in previous section

Added a link to ingress in d70f94bef -> 65cbc79e3.

File: docs/user/aws/install_upi.md (hunk 3 above)
Comment (wking) [reply docs/user/aws/install_upi.md]: > I think we were also going to list the required subdomains...

Oh, right.  I've added a FIXME and restored the WIP with d70f94bef -> 65cbc79e3 while I hunt those down.

**Pull request merged**